pilot/pkg/proxy/envoy/v2/gen2.go

```diff
@@ -0,0 +1,165 @@
+		w = &model.WatchedResource{
+			TypeUrl: t,
+		}
+		con.node.Active[t] = w
+		isAck = false // newly watched resource
+	}
+	con.mu.RUnlock()
+
+	if discReq.ErrorDetail != nil {
+		errCode := codes.Code(discReq.ErrorDetail.Code)
+		adsLog.Warnf("ADS: ACK ERROR %s %s:%s", con.ConID, errCode.String(), discReq.ErrorDetail.GetMessage())
+		return w, true
+	}
+
+	if discReq.ResponseNonce == "" {
+		isAck = false // initial request
+	}
+	// This is an ACK response to a previous message - but it may refer to a response on a previous connection to
+	// a different XDS server instance.
+	nonceSent := w.NonceSent
+
+	// GRPC doesn't send version info in NACKs for RDS. Technically if nonce matches
+	// previous response, it is an ACK/NACK.
+	if nonceSent != "" && nonceSent == discReq.ResponseNonce {
+		adsLog.Debugf("ADS: ACK %s %s %s %v", con.ConID, discReq.VersionInfo, discReq.ResponseNonce,
+			time.Since(w.LastSent))
+		w.NonceAcked = discReq.ResponseNonce
+	}
+
+	if nonceSent != discReq.ResponseNonce {
+		adsLog.Debugf("ADS:RDS: Expired nonce received %s, sent %s, received %s",
+			con.ConID, nonceSent, discReq.ResponseNonce)
+		rdsExpiredNonce.Increment()
+		// This is an ACK for a resource sent on an older stream, or out of sync.
+		// Send a response back.
+		isAck = false
+	}
+
+	// Change in the set of watched resource - regardless of ack, send new data.
+	if !listEqualUnordered(w.ResourceNames, discReq.ResourceNames) {
+		isAck = false
+		w.ResourceNames = discReq.ResourceNames
+	}
+
+	return w, isAck
+}
+
+// handleCustomGenerator uses model.Generator to generate the response.
+func (s *DiscoveryServer) handleCustomGenerator(con *XdsConnection, req *xdsapi.DiscoveryRequest) error {
+	w, isAck := s.handleReqAck(con, req)
+	if isAck {
+		return nil
+	}
+
+	push := s.globalPushContext()
+	resp := &xdsapi.DiscoveryResponse{
+		TypeUrl:     w.TypeUrl,
+		VersionInfo: push.Version, // TODO: we can now generate per-type version !
+		Nonce:       nonce(push.Version),
+	}
+
+	cl := con.node.XdsResourceGenerator.Generate(con.node, push, w)
+	sz := 0
+	for _, rc := range cl {
+		resp.Resources = append(resp.Resources, rc)
+		sz += len(rc.Value)
+	}
+
+	err := con.send(resp)
+	if err != nil {
+		adsLog.Warnf("ADS: Send failure %s: %v", con.ConID, err)
+		recordSendError(apiSendErrPushes, err)
+		return err
+	}
+	apiPushes.Increment()
+	w.LastSent = time.Now()
+	w.LastSize = sz // just resource size - doesn't include header and types
+	w.NonceSent = resp.Nonce
+
+	return nil
+}
+
+// TODO: verify that ProxyNeedsPush works correctly for Generator - ie. Sidecar visibility
+// is respected for arbitrary resource types.
+
+// Called for config updates.
+// Will not be called if ProxyNeedsPush returns false - ie. if the update
+func (s *DiscoveryServer) pushGeneratorV2(con *XdsConnection, push *model.PushContext, currentVersion string, w *model.WatchedResource) error {
+	// TODO: generators may send incremental changes if both sides agree on the protocol.
```

Comment:
Seems like we are trying to invent a new protocol here... if XDS isn't sufficient we should try to make changes there?

Reply:
XDS is sufficient actually - EDS is doing this already ( sending only clusters that have changes).

It happens that LDS and CDS expect the full list in one message - but that's an implementation of the resource, not the protocol fault. 

XDS even allow mixing different resource types in a response - so we have everything we need.

Reply:
I don't think you can mix: See https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/discovery.proto#envoy-api-msg-discoveryresponse
"must be consistent with the type_url in the ‘resources’ repeated Any (if non-empty)."

Reply:
Oh maybe if you make the type_url empty there?

Reply:
I can make it empty. There is a proposal - https://github.com/envoyproxy/envoy/pull/10689 - that seems to allow mixed responses, if I read it correctly. The use case is pretty important - gRPC only needs the endpoints, but needs to do 4 roundtrips to get it. The gRPC generator could return all 4 types of response - knowing that the client will need them and can process them. 

I would also argue the response is consistent - the empty resource is used for that type.

I can also use the real type_url and send an empty MCP resource, or one with a special name.
But simpler is better.